docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,279 @@
+order is hard-coded. We would therefore deprecate the `--admission-control`
+setting from `pkg/genericapiserver/options.ServerRunOptions`.
+
+
+## One config file per plugin
+
+All admission plugins currently share a single config file, specified via
+`--admission-control-config-file`. This is problematic if multiple plugins
+decide they want to use the same names for their configuration sections. For
+example, if plugin `Foo` wants
+
+```yaml
+myConfig:
+  someField: red
+```
+
+and plugin `Bar` wants
+
+```yaml
+myConfig:
+  someField:
+  - one
+  - two
+  - three
+```
+
+this won't work because of the naming collision and the type of `someField` is
+different for the two plugins. What we'd like to see instead is:
+
+1. deprecate `--admission-control-config-file`
+1. replace it with `--admission-plugin PLUGIN=CONFIG_FILE` (or an equivalent
+   setup in a structured configuration, once we've fully moved over)
+
+You would specify `--admission-plugin` once per plugin that you want to
+configure.
+
+
+## Ability to enable/disable individual plugins
+
+We want to hard-code a set of plugins that are enabled by default.
+Administrators need to be able to override the defaults, of course. To simply
+enable or disable a specific plugin, you would create a config file that looks
+like this:
+
+```yaml
+kind: DefaultAdmissionConfig
+version: v1
+enabled: false
+```
+
+The flow to determine if a plugin is enabled would look like this:
+
+1. If the plugin is on by default
+  1. If there is no config file for it, it remains enabled
+  1. Otherwise (have config file)
+    1. If it deserializes to `DefaultAdmissionConfig`, it uses the value of
+       `enabled` from the config file
+    1. Otherwise, it is up to the individual plugin to deserialize its config
+       and determine if it should be enabled
+1. Otherwise (off by default)
+  1. Same logic as above, except if there is no config file, it remains disabled
+
+
+## Refine the interface
+
+The current interface looks like this:
+
+```go
+type Interface interface {
+  Admit(a Attributes) (err error)
+  Handles(operation Operation) bool
+}
+```
+
+We'd like to break this into multiple phases, called at the appropriate times
+within the request lifecycle.
+
+```go
+type Interface interface {
+	// ApplyDefaults is executed after the request body is deserialized into an
+	// object for create and update requests. Plugins are free to modify the
+	// object as they see fit. If an error is returned, the request is marked as
+	// failed and no changes are persisted.
+	ApplyDefaults(a Attributes) error
+
+	// Authorize is executed after all plugins' ApplyDefaults methods have
+	// executed. This gives plugins a chance to decide if the request should be
+	// allowed to proceed after the object has had all of its defaults applied. If
+	// an error is returned, the request is marked as failed and no changes are
+	// persisted.
+	Authorize(a Attributes) error
```

Comment:
This proposal is advocating for admission hooks post-validation.

Given that structure, admission is in two phases:

1. mutating the object to set fields (must be done post validation)
2. observing the object to enforce constraints (seems sensible to always do post validation)

@deads2k - if your advocating that it's not worth doing any admission after validation, then that is a different topic altogether.  If we do have a hook after validation, I think it makes sense to just have admission be in two phases of pre/post validation, and have post-validation always get a read-only view of the object passed in.  i view not having this split as an error worth fixing independent of any quota scenario.